### ISSUE TITLE
Folder for external figures

### ISSUE THREAD

**Issue by xiangzhu** [workflowr/workflowr#37]
Hi John,

I am curious whether `workflowr` framework defines a specific folder for users to put external figures. Can users define their own folders such as `figure` for this purpose?

Examples of using external figures:

1. When I write teaching materials, sometimes I need to use figures published in other people's paper.

2. When showing some technical details, I often take a screen shot of pdf files complied from latex files, instead of coping the latex chunks.

**Comment by pcarbo** [comment]
I think that is an excellent suggestion. You could also think about linking to other files (e.g., a paper in PDF format). How Raman did this for the Stephens lab website was to include these sorts of files in an `assets` folder.

**Comment by xiangzhu** [comment]
@pcarbo  yes, `assets` seems to be a better name if this folder aims for hosting all external files (PDF, PNG, etc).

@jdblischak  I think that a more general question is whether creating these folders (`figure`, `assets`, etc) will do any harm to the current `workflowr` framework?

**Comment by xiangzhu** [comment]
Here is another issue of including external figures (in contrast to the figures generated by `Rmd` files).

I first manually created a folder `assets` for external figures:
![screenshot from 2017-04-17 14-38-44](https://cloud.githubusercontent.com/assets/5986023/25101873/94c7434c-237b-11e7-9b27-205bc2f38414.png)

If I add external figures as follows, the figures will **not** be shown on website:
```Rmd
![](../assets/gwas_info.png)
```

I guess this is because the rendered `html` uses a local path for the figure:
```html
<img src="/home/xiangzhu/Github/replication-time/assets/gwas_info.png" />
```

To address this issue, I have to write `Rmd` file as follows:
```Rmd
![](https://raw.githubusercontent.com/xiangzhu/replication-time/master/assets/gwas_info.png)
```

Then the resulting `html` will use the correct path:
```html
<img src="https://raw.githubusercontent.com/xiangzhu/replication-time/master/assets/gwas_info.png" />
```

Related commits:

- https://github.com/xiangzhu/replication-time/commit/56b44a950d64ef966a2767ec688d0adfff87504b
- https://github.com/xiangzhu/replication-time/commit/b22ec8475803b9c6c11c3a8d9c8c9c5dfb938ace

**Comment by jdblischak** [comment]
> I think that a more general question is whether creating these folders (figure, assets, etc) will do any harm to the current workflowr framework?

@xiangzhu Creating extra folders will not harm the workflowr framework so long as the folders do not use the names already used by workflowr.

Including custom images is definitely allowed, but it requires some manual intervention. The main constraint is that only content that is present inside `docs/`will be served on the website. Your solution gets around this problem by specifying the full path to the file on GitHub. This approach has some downsides. First, when prototyping, the figure will not show up in your Rmd until after you push the figure to GitHub. Second, the path will always be to the image in your repository, even if someone else forks the repo and changes the figure. Third, it's a lot of typing.

Here are two suggestions:

1. To use an `assets` folder that holds all the external images and other files, you could move this folder to be within `docs/`: `git mv assests/ docs/assets`. Then you could specify the path to the image with `![](assets/gwas_info.png)`

2. If you will have lots of custom images, you could organize them using the same organization as the automatically generated figures. The figure images for the file `rtQTL_20170314.Rmd` are saved in `docs/figure/rtQTL_20170314.Rmd`, so you could save the custom image there: `git mv assets/gwas_info.png docs/figure/rtQTL_20170314.Rmd`. Then the path to the figure would be specified with `![](figure/rtQTL_20170314.Rmd/gwas_info.png)`

What do you think of those potential solutions?

**Comment by xiangzhu** [comment]
Personally I think option 1 is better, at least for my current case.

I just make the change and it is working! Many thanks!

**Comment by jdblischak** [comment]
Following up on my previous suggestions for including external images, the second suggestion of putting the image files in the same directory as the ones generated from the R Markdown file is no longer a valid solution. Now workflowr deletes the figures in that directory prior to building the HTML, in order to avoid keeping around unnecessary files.

Thus my recommendation is the following:

1. Create an assets directory to include any file that should be part of the website but is not created by the R Markdown files in `analysis/`:
    ```
   dir.create("docs/assets")
    ```
1. Move the image files to `docs/assets`
1. In the R Markdown file, refer to these images using the relative path from `docs/` (because this is where the HTML files are located):
    ```
   ![](assets/external.png)
    ```
1. Run `wflow_build()` to confirm the external image was properly added
1. Use `wflow_git_commit()` to commit the files to the Git repo (so that they get pushed to GitHub)
    ```
   wflow_git_commit("docs/assets/external.png", "Add external image of ...")
   # If you are adding multiple files, you could use a file glob
   wflow_git_commit("docs/assets/*.png", "Add external images of ...")
    ```
1. Run `wflow_publish()` on the R Markdown file that contains the external image(s)

**Comment by ryurko** [comment]
Just a minor comment / request to indicate in the FAQ to use 'public/assets' instead of 'docs/assets' with GitLab Pages.

**Comment by jdblischak** [comment]
> Just a minor comment / request to indicate in the FAQ to use 'public/assets' instead of 'docs/assets' with GitLab Pages.

@ryurko Thanks for the suggestion! Fixed in commit bdffc17a6113fdc2254d3f6b317cd0999c832db4. The documentation has the updated text:

https://jdblischak.github.io/workflowr/articles/wflow-05-faq.html#how-can-i-include-external-images-in-my-website
